api/src/string_array.c

```diff
@@ -26,31 +26,95 @@
+
+    if (NulledStrings != NULL)
+    {
+        result = (AwaStringArray *)AwaArray_New();
+        AwaArray_SetContext((AwaArray *)result, NulledStrings);
+    }
+
+    return result;
 }
 
 void AwaStringArray_Free(AwaStringArray ** array)
 {
-    AwaArray_Free((AwaArray **)array, AwaResourceType_StringArray);
+    if ((array != NULL) && (*array != NULL))
+    {
+        AwaArray * NulledStrings = AwaArray_GetContext((AwaArray *)*array);
+        if (NulledStrings != NULL)
+        {
+            AwaArray_Free(&NulledStrings, AwaResourceType_StringArray);
+        }
+        AwaArray_Free((AwaArray **)array, AwaResourceType_StringArray);
+    }
 }
 
 void AwaStringArray_SetValueAsCString(AwaStringArray * array, AwaArrayIndex index, const char * value)
 {
-    Array_SetValue((AwaArray *)array, index, (void *)value, strlen(value) + 1);
+    if (array != NULL)
+    {
+        AwaArray * NulledStrings = AwaArray_GetContext((AwaArray *)array);
+        if (NulledStrings != NULL)
+        {
+            Array_SetValue((AwaArray *)NulledStrings, index, (void *)value, strlen(value) + 1);
+        }
+        Array_SetValue((AwaArray *)array, index, (void *)value, strlen(value));
+    }
 }
 
 void AwaStringArray_DeleteValue(AwaStringArray * array, AwaArrayIndex index)
 {
-    Array_DeleteItem((AwaArray *)array, index, AwaResourceType_StringArray);
+    if (array != NULL)
+    {
+        AwaArray * NulledStrings = AwaArray_GetContext((AwaArray *)array);
+        if (NulledStrings != NULL)
+        {
+            Array_DeleteItem((AwaArray *)NulledStrings, index, AwaResourceType_StringArray);
+        }
+        Array_DeleteItem((AwaArray *)array, index, AwaResourceType_StringArray);
+    }
 }
 
 const char * AwaStringArray_GetValueAsCString(const AwaStringArray * array, AwaArrayIndex index)
 {
-    return Array_GetValue((const AwaArray *)array, index);
+    char * value = NULL;
+
+    if (array != NULL)
+    {
+        AwaArray * NulledStrings = AwaArray_GetContext((AwaArray *)array);
+        if (NulledStrings != NULL)
+        {
+            value = Array_GetValue((const AwaArray *)NulledStrings, index);
+
+            if (value == NULL)
+            {
+                const char * nonNulledValue = Array_GetValue((AwaArray *)array, index);
+                int nonNulledLength = Array_GetValueLength((AwaArray *)array, index);
+                char *  nulledValue = (char *)malloc(nonNulledLength + 1);
+
+                if ((nonNulledValue != NULL) && (nulledValue != NULL))
+                {
+                    memcpy(nulledValue, nonNulledValue, nonNulledLength);
+                    nulledValue[nonNulledLength] = '\0';
+                    Array_SetValue(NulledStrings, index, nulledValue, nonNulledLength + 1);
+                    free(nulledValue);
+                    value = Array_GetValue((const AwaArray *)NulledStrings, index);
+                }
+                else
+                {
+                    free(nulledValue);
```

Comment:
I'm a little surprised by this - which platforms is this broken on? The C99 standard requires it to be a no action: http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf section 7.20.3.2.